compiler/src/iree/compiler/Dialect/Flow/Transforms/RaiseSpecialOps.cpp

```diff
@@ -27,6 +28,170 @@ namespace Flow {
 
 namespace {
 
+static LogicalResult raiseTensorExtractToInput(linalg::GenericOp linalgOp,
+                                               RewriterBase &rewriter) {
+  if (!linalgOp.hasTensorSemantics())
+    return failure();
+
+  // Find a tensor.extract op in the linalgOp body.
+  tensor::ExtractOp extractOp;
+  linalgOp.walk([&](tensor::ExtractOp op) {
+    if (extractOp)
+      return WalkResult::interrupt();
+    extractOp = op;
+    return WalkResult::advance();
+  });
+
+  if (!extractOp)
+    return failure();
+
+  // Raise the tensor.extract op to an input.
+  SmallVector<AffineExpr> exprs;
+  for (Value indexValue : extractOp.getIndices()) {
+    // For raising, the indexing value must be one of the following:
+    //    1. A constant value.
+    //    2. A linalg.index.
+
+    // 1. Indexing value is a constant.
+    APInt constantIndex;
+    if (matchPattern(indexValue, m_ConstantInt(&constantIndex))) {
+      exprs.push_back(getAffineConstantExpr(constantIndex.getLimitedValue(),
+                                            rewriter.getContext()));
+      continue;
+    }
+    // 2. The indexing value is a linalg.index.
+    assert(isElementwise(linalgOp) && "Expected elementwise linalg op");
+    if (auto indexOp = indexValue.getDefiningOp<linalg::IndexOp>()) {
+      exprs.push_back(
+          getAffineDimExpr(indexOp.getDim(), rewriter.getContext()));
+      continue;
+    }
+    return failure();
+  }
+  AffineMap indexingMap = AffineMap::get(
+      /*dimCount=*/linalgOp.getNumLoops(),
+      /*symbolCount=*/0, exprs, rewriter.getContext());
+
+  // Replace the linalgOp with a new linalgOp where the source tensor is
+  // an input with the indexing map.
+  SmallVector<Value> newInputs = linalgOp.getInputs();
+  newInputs.insert(newInputs.begin(), extractOp.getTensor());
```

Comment:
why `insert`, and not `push_back`?

Reply:
Adding the new input as the last input adds a new block argument between the input/init block arguments. It's easier to create the IR mapping if there is a new argument at start.